nasl/nasl_crypt_helper.c

```diff
@@ -0,0 +1,197 @@
+#include "../misc//support.h"
+#include "nasl_crypto_helper.h"
+#include "nasl_debug.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <gcrypt.h>
+#include <glib.h>
+#include <gvm/base/logging.h>
+#include <stddef.h>
+#include <stdlib.h>
+
+void *
+hmac_md5_for_prf (const void *key, int keylen, const void *buf, int buflen)
+{
+  void *signature = g_malloc0 (16);
+  gsize signlen = 16;
+  GHmac *hmac;
+
+  hmac = g_hmac_new (G_CHECKSUM_MD5, key, keylen);
+  g_hmac_update (hmac, buf, buflen);
+  g_hmac_get_digest (hmac, signature, &signlen);
+  g_hmac_unref (hmac);
+  return signature;
+}
+
+void *
+hmac_sha1 (const void *key, int keylen, const void *buf, int buflen)
+{
+  void *signature = g_malloc0 (20);
+  gsize signlen = 20;
+  GHmac *hmac;
+
+  hmac = g_hmac_new (G_CHECKSUM_SHA1, key, keylen);
+  g_hmac_update (hmac, buf, buflen);
+  g_hmac_get_digest (hmac, signature, &signlen);
+  g_hmac_unref (hmac);
+  return signature;
+}
+
+void *
+hmac_sha256 (const void *key, int keylen, const void *buf, int buflen)
+{
+  void *signature = g_malloc0 (32);
+  gsize signlen = 32;
+  GHmac *hmac;
+
+  hmac = g_hmac_new (G_CHECKSUM_SHA256, key, keylen);
+  g_hmac_update (hmac, buf, buflen);
+  g_hmac_get_digest (hmac, signature, &signlen);
+  g_hmac_unref (hmac);
+  return signature;
+}
+
+void *
+hmac_sha384 (const void *key, int keylen, const void *buf, int buflen)
+{
+  gcry_md_hd_t hd;
+  gcry_error_t err;
+  void *ret;
+
+  if (!buf || buflen <= 0)
+    return NULL;
+
+  err = gcry_md_open (&hd, GCRY_MD_SHA384, key ? GCRY_MD_FLAG_HMAC : 0);
+  if (err)
+    {
+      g_message ("nasl_gcrypt_hash(): gcry_md_open failed: %s/%s",
+                 gcry_strsource (err), gcry_strerror (err));
+      return NULL;
+    }
+
+  if (key)
+    {
+      err = gcry_md_setkey (hd, key, keylen);
+      if (err)
+        {
+          g_message ("nasl_gcrypt_hash(): gcry_md_setkey failed: %s/%s",
+                     gcry_strsource (err), gcry_strerror (err));
+          return NULL;
+        }
+    }
+
+  gcry_md_write (hd, buf, buflen);
+  ret = g_memdup2 (gcry_md_read (hd, 0), 48);
+  gcry_md_close (hd);
+  return ret;
+}
+
+gpg_err_code_t
+mac (const char *key, const size_t key_len, const char *data,
+     const size_t data_len, const char *iv, const size_t iv_len, int algo,
+     int flags, char **out, size_t *out_len)
+{
+  // guardian
+  gcry_err_code_t result = 0;
+  gcry_mac_hd_t hd;
+  if (key == NULL || key_len < 1)
+    return GPG_ERR_MISSING_KEY;
+  if (data == NULL || data_len < 1)
+    return GPG_ERR_MISSING_VALUE;
+  if (out == NULL)
+    {
+      return GPG_ERR_GENERAL;
+    }
```

Comment:
I am a bit confused about the error codes used here. The return type is `gpg_err_code_t`, the type of `result` is `gcry_err_code_t`, the codes returned, when a key is missing is again some GPG stuff and the actual `gcry_mac` functions return a `gcry_error_t`

Reply:
Those are the same:
```
typedef gpg_error_t gcry_error_t;
```
but you're right I will change it :).